rclcpp/src/rclcpp/clock.cpp

```diff
@@ -182,6 +182,62 @@ Clock::sleep_for(Duration rel_time, Context::SharedPtr context)
   return sleep_until(now() + rel_time, context);
 }
 
+bool
+Clock::is_valid()
+{
+  switch (get_clock_type()) {
+    case RCL_ROS_TIME:
+    case RCL_STEADY_TIME:
+    case RCL_SYSTEM_TIME:
+      return now().nanoseconds() > 0;
+
+    case RCL_CLOCK_UNINITIALIZED:
+    default:
+      return false;
+  }
+}
+
+bool
+Clock::wait_for_valid(Context::SharedPtr context)
+{
+  if (get_clock_type() == RCL_CLOCK_UNINITIALIZED) {
+    throw std::runtime_error(
+            "clock cannot be waited on as its clock_type is RCL_CLOCK_UNINITIALIZED");
+  }
+
+  // Wait until the first valid time
+  return sleep_until(rclcpp::Time(0, 1, get_clock_type()), context);
```

Comment:
Why do we need to call this? if this is the definition of valid or invalid, this should be called in `is_valid` method? this method is meant to loop forever to check if it becomes into valid?

Reply:
If a clock is reflecting time 0, it is invalid. This waits until the clock becomes valid.

This is particularly useful if sim_time is active, since a time 0 will be invalid, and the simulation can remain paused.
In which case, the clock will wait until the simulation resumes and time moves forward into a valid, non-zero time